### DISCUSSION TITLE
Livewire page not subcribing to listening channels

### DISCUSSION THREAD

**Discussion by judgej** [livewire/livewire#6347] [1. Help]
I'm cross-posting this in the hope that someone can direct me at something I am missing. Been stumped on this all weekend.

First posted here: https://laracasts.com/discuss/channels/livewire/livewire-page-not-subscribing-to-websockets-channels

Running local sail, Laravel 9, Livewire 2, beyondcode/laravel-websockets 1.14

The websockets server (`php artisan websockets:serve`) runs and shows connections, subscriptions and events. I can see these also in the websockets test panel http://localhost:8040/laravel-websockets

My simple livewire page connects to websockets, and I see that too on a browser refresh:

![image](https://github.com/livewire/livewire/assets/395934/788b2925-03f8-431c-9561-aea0150c8be1)

What I don't see is any subscriptions by that web page. I'm assuming that if the livewire page is listening to channels, then it needs to subscribe, and those subscriptions will appear in the websockets log (see through the websockets panel).

The Livewire page is listening like this:

```php
    public function getListeners(): array
    {
        return [
            sprintf('echo:order.%s,.order.updated', $this->orderUuid) => 'handleOrderUpdated',
            // sprintf('echo:order.%s,order.updated', $this->orderUuid) => 'handleOrderUpdated',
            // sprintf('echo:order.%s,App\Events\OrderUpdated', $this->orderUuid) => 'handleOrderUpdated',
        ];
    }
```

for example:

```php
array:1 [▼ // resources/views/livewire/test.blade.php
  "echo:order.106fc4ef-94fe-4f1d-bcf8-1844e69a0a46,.order.updated" => "handleOrderUpdated"
]
```

I can broadcast this event (using `ShouldBroadcastNow` for now) and can see the event in the websockets panel:

    Channel: order.106fc4ef-94fe-4f1d-bcf8-1844e69a0a46, Event: order.updated

The broadcast is being broadcast, so far as I can see. But the Livewire page simply does not see it. I'm assuming it does not see it because it did not subscribe.

I realise that there is probably a tonne of additional information needed to diagnose this, but I'm posting this on the off-chance that somebody else has had the same problem and can joins a ferw dots.

Just to note: I can see no errors anywhere, on the command line, browser console, HTTP requests and responses. Everything is behaving as I would expect, except the Livewire page is not subscribing to the public channel.

Any ideas?

**Comment by judgej** [comment]
The subscription to a channel - is that the front end asking to subscribe to a channel through the websocket directly, or does Laravel subscribe at the back end in behalf of the front end? Just want to make sure my assumption of the former approach is correct.

Edit: nice explanation here - https://stackoverflow.com/a/64049329 The Laravel back end, if I understand correctly, creates a JWT for the private channels the front end is allowed to subscribe to. This JWT is then used by the front end as authorisation when subscibing. It's a JWT, so the pusher server does not need to go back to your app to check authorisation - so long as it has the same secret as your app, then the JWT is enough to give your browser access to the [private] channels.

So, in my app, I think I can see the JWT going to the front end when the Livewire component (one for each component, I guess) is first mounted. As an aside, this explains why channel listeners are not dynamic, why they are frozen when the component is mounted. So I think in my app something is going wrong at the front end. The laravel echo library is not doing the subscribing, or perhaps is not properly being asked to do the subscribing to the pusher channel.

**Reply by thannaske** [reply]
Did you ever solve the issue? I have the exact same problem and I can't figure out what I'm doing wrong, I basically just copy-pasted the documentation's examples and Echo is connected to the Websocket server but does never subscribe to the channel.

**Reply by judgej** [reply]
I didn't, no. It is a very difficult area to navigate because IMO the documentation is very unclear. There are config items that are used by the websockets server, others by the laravel backend, yet others by the browser front end, and some that are used in a mish-mash of places, so it it difficult to know what will be affected by each change.

The conclusion I've come to, is to keep the websockets server and the web application as two completely separate systems, and not to try to get a laravel app to serve its own websockets. So effectively you would be running your own version of pusher, and that could serve as many sites as you like (another benefit when setting up sites - single point of failure I know, but also a single poing of robust configuration). Not quite got there yet, and there are alternatives to using laravel to serve the pusher service (which are probably getting more support anyway). There is no reason why the pusher service even needs to be PHP application - we are only interested in the HTTP and websockets that it serves, with some authentication credentials thrown in.

Sorry - turned into a semi-rant, which isn't helpful - but maybe it will help if it avoids fighting a laravel websockets plugin that isn't getting a lot of care these days from fom the people who really understand how it works. And that's no criticism of the excellent work that went into it. Complex things can be time consuming.
